### PR TITLE
[BACKLOG-40064] - VFS Connections -  Non-admin users would only see the connections list when on the Open/Save dialogs

### PR DESCRIPTION
**Pull request by soagarwal1**
@pentaho/hoth Please review

### CODE REVIEW

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java [hunk 1]
```diff
@@ -188,6 +190,11 @@ private <T> void render( TreeNode treeNode, T tree ) {
     TreeItem childTreeItem = createTreeItem( treeNode, tree );
     if ( treeNode.hasChildren() ) {
       for ( TreeNode childTreeNode : treeNode.getChildren() ) {
+        Repository repository = Spoon.getInstance().getRepository();
+        if( childTreeNode.getLabel().equals( "VFS Connections" ) && repository!=null
```
Comment (dcleao) [inline ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:194]: Generally, it seems that this is being done after the fact: that we should be trying to _prevent_ whoever is adding the VFS Connections node to the tree from doing so, instead of "harcoding" logic to _remove_ a specific tree node in this central location. AFAICT, up until now, the tree manager has no specific knowledge of VFS Connections or other specific nodes.

Otherwise, if this is the only possible location to intervene, see the following smaller issues:
1. Check for `repository != null` first.
2. Check the tree node using some identifier, and not the label, given that will be locale dependent.
3. Formatting is off.

**Review by dcleao — CHANGES_REQUESTED** [review]

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 31m 46s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -Djs.no.sandbox -pl ui
```
---
:ok_hand: All tests passed!
 
Tests run: **1434**, Failures: **0**, Skipped: **6** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9219/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java [hunk 2]
```diff
@@ -188,6 +190,11 @@ private <T> void render( TreeNode treeNode, T tree ) {
     TreeItem childTreeItem = createTreeItem( treeNode, tree );
     if ( treeNode.hasChildren() ) {
       for ( TreeNode childTreeNode : treeNode.getChildren() ) {
+        Repository repository = Spoon.getInstance().getRepository();
```
Comment (NJtwentyone) [inline ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:193]: I don't like the introduction `Spoon` and `Repository `concepts here in `TreeManager.java`. 

The function declaration using generics supports this

``` private <T> void render( TreeNode treeNode, T tree ) ```

I think in the scenarios of:

- proof of concept
- release pressures 

This would be "acceptable". However, in my opinion this is the beginning of the release.

`TreeManager.java` seems pretty generic and the classes `TreeNode.java` used in render() are also pretty generic.

Possible course of actions.
You should figure out how `rootNodes` is populated and apply this filter or block, earlier in the process instead of here in `TreeManager#render`
```
  public void render() {
    rootNodes.forEach( t -> {
      render( t, tree );
    } );
    setExpanded();
  }
```
  
  https://github.com/pentaho/pentaho-kettle/blob/4adc94b022be5d48323af8dbda8db171dd21495f/ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java#L94-L99

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
I don't feel this the right place to add this code see comment: https://github.com/pentaho/pentaho-kettle/pull/9219/files#r1517956769

However, not technically incorrect, I can be overruled. If you can find two other people to get their explicit approval you can merge per general pentaho PR rules.

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java [hunk 3]
```diff
@@ -188,6 +190,11 @@ private <T> void render( TreeNode treeNode, T tree ) {
...
+        if( childTreeNode.getLabel().equals( "VFS Connections" ) && repository!=null
+          && repository.getUserInfo()!=null && !repository.getUserInfo().isAdmin() ){
+          continue;
+        }
```
Comment (NJtwentyone) [inline ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:197]: Most likely the change will be in java, so a unit test will be in order.

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 1 above)
Comment (NJtwentyone) [inline ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:194]: Would checking  for an english label work in other internalization scenarios (i18n) ?

`childTreeNode.getLabel().equals( "VFS Connections" )`

Please check and verify .

Comment (NJtwentyone) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:194]: #2 i made similar observation with this comment - https://github.com/pentaho/pentaho-kettle/pull/9219#discussion_r1517976933.

But that might be irrelevant if TreeNode "VFS Connections" is never added to `rootNodes` as I commented in https://github.com/pentaho/pentaho-kettle/pull/9219#discussion_r1517956769

Comment (NJtwentyone) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java:194]: duplicate observation for item-2 from comment: https://github.com/pentaho/pentaho-kettle/pull/9219#discussion_r1517591573

Comment (soagarwal1) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: Changed the location of code, please re-review

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 2 above)
Comment (soagarwal1) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: Changed the location of code, please re-review

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 1 above)
Comment (soagarwal1) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: Changed the location of code, please re-review

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 3 above)
Comment (soagarwal1) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: holding on to the test case till re-review is complete for acceptance of location of code

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 22m 16s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -Djs.no.sandbox -pl \
plugins/connections/ui
```
---
 :exclamation: No tests found!
 ---
 
 

:information_source:  _This is an automatic message_

File: plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java [hunk 4]
```diff
@@ -58,4 +60,14 @@ public void refresh( AbstractMeta meta, TreeNode treeNode, String filter ) {
   public String getTitle() {
     return STRING_VFS_CONNECTIONS;
   }
+
+  @Override
+  public void create( AbstractMeta meta, TreeNode parent ) {
```
Comment (NJtwentyone) [inline plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java:65]: I like this new placement of the code.

File: plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java [hunk 5]
```diff
@@ -58,4 +60,14 @@ public void refresh( AbstractMeta meta, TreeNode treeNode, String filter ) {
...
+    Repository repository = Spoon.getInstance().getRepository();
```
Comment (NJtwentyone) [inline plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java:66]: You'll probably get a check style for this `repository!=null `--> `repository != null`

File: plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java [hunk 6]
```diff
@@ -58,4 +60,14 @@ public void refresh( AbstractMeta meta, TreeNode treeNode, String filter ) {
...
+    if( repository != null && repository.getUserInfo() != null && !repository.getUserInfo().isAdmin() ) {
```
Comment (NJtwentyone) [inline plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java:67]: Add unit test for ConnectionFolderProvider.java#create. 

It should cover 

- when a user is not connected to a repository
- when a user is connected to a repository and is not Admin
- when a user is connected to a repository and isAdmin

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
Good placement of the code. I would now be looking for unit tests see: https://github.com/pentaho/pentaho-kettle/pull/9219/files#r1521550377

Unless there is some major effort due to static classes and mocking. If it does I can probably help.

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 2 above)
Comment (NJtwentyone) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: I agree with new location, it looks good.

File: ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java (hunk 1 above)
Comment (NJtwentyone) [reply ui/src/main/java/org/pentaho/di/ui/spoon/tree/TreeManager.java]: moved to ConnectionFolderProvider, looks good.

**Review by dcleao — APPROVED** [review]
All good, except it is missing unit tests, as noted by @NJtwentyone, in https://github.com/pentaho/pentaho-kettle/pull/9219/files#r1521550377.

File: plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java (hunk 6 above)
Comment (soagarwal1) [reply plugins/connections/ui/src/main/java/org/pentaho/di/connections/ui/tree/ConnectionFolderProvider.java:67]: Added Unit tests

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 29m 2s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -Djs.no.sandbox -pl \
plugins/connections/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **8**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9219/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
You do have a code smell.

the function `!repository.getUserInfo().isAdmin()` returns type `**Boolean**`. so you need to check for null and then check that is then true/false
Screenshot 2024-03-14 at 10 47 08 AM

And then you need to update your unit tests to handle this `Boolean `return value being null. (the original author should have just used a primitive boolean)

You can run sonarQube locally via SonarLint: https://hv-eng.atlassian.net/wiki/spaces/~nicholas.jordan/pages/4265148420/SonarLint+IntelliJ+Integration

**Comment by soagarwal1** [comment]
Resolved Unit test case and code smell issues

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 23m 21s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -Djs.no.sandbox -pl \
plugins/connections/ui
```
---
:ok_hand: All tests passed!
 
Tests run: **10**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9219/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by NJtwentyone** [comment]
code coverage is high, I'm not sure why it is registering 0.0% that is an issue outside of this PR

**Review by NJtwentyone — APPROVED** [review]
